pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,431 @@
+func (n *nodeLogQuery) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int) {
+	cmdStr, args, err := getLoggingCmd(n, services)
+	if err != nil {
+		fmt.Fprintf(w, "\nfailed to get logging cmd: %v\n", err)
+		return
+	}
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "\nerror: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		fmt.Fprintf(w, "\ncontext error: %v\n", ctx.Err())
+		return
+	}
+
+	for _, service := range services {
+		heuristicsCopyFileLogs(ctx, w, service)
+	}
+}
+
+// heuristicsCopyFileLogs attempts to collect logs from either
+// /var/log/service
+// /var/log/service.log or
+// /var/log/service/service.log or
+// in that order stopping on first success.
+func heuristicsCopyFileLogs(ctx context.Context, w io.Writer, service string) {
+	logFiles := [7]string{
+		filepath.Join(logDir, fmt.Sprintf("%s", service)),
+		filepath.Join(logDir, fmt.Sprintf("%s.log", service)),
+		filepath.Join(logDir, service, fmt.Sprintf("%s.log", service)),
+	}
+
+	var err error
+	for _, logFile := range logFiles {
+		err = heuristicsCopyFileLog(ctx, w, logFile)
+		if err == nil {
+			break
+		} else if errors.Is(err, os.ErrNotExist) {
+			continue
+		} else {
+			break
+		}
+	}
+
+	if err != nil {
+		// If the last error was file not found it implies that no log file was for found for the service
+		if errors.Is(err, os.ErrNotExist) {
+			fmt.Fprintf(w, "\nlog not found for %s\n", service)
+			return
+		}
+		fmt.Fprintf(w, "\nerror getting log for %s: %v\n", service, err)
+	}
+}
+
+// readerCtx is the interface that wraps io.Reader with a context
+type readerCtx struct {
+	ctx context.Context
+	io.Reader
+}
+
+func (r *readerCtx) Read(p []byte) (n int, err error) {
+	if err := r.ctx.Err(); err != nil {
+		return 0, err
+	}
+	return r.Reader.Read(p)
+}
+
+// newReaderCtx gets a context-aware io.Reader
+func newReaderCtx(ctx context.Context, r io.Reader) io.Reader {
+	return &readerCtx{
+		ctx:    ctx,
+		Reader: r,
+	}
+}
+
+// heuristicsCopyFileLog returns the contents of the given logFile
+func heuristicsCopyFileLog(ctx context.Context, w io.Writer, logFile string) error {
+	fInfo, err := os.Stat(logFile)
+	if err != nil {
+		return err
+	}
+	// This is to account for the heuristics where logs for service foo
+	// could be in /var/log/foo/
+	if fInfo.IsDir() {
+		return os.ErrNotExist
+	}
+
+	f, err := os.Open(logFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(w, newReaderCtx(ctx, f)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func safeString(s string) error {
+	if len(s) > 100 {
```

Comment:
Where does the 100 limit come from? Probably worth adding a comment. The systemd limit for a unit name is 256 characters https://www.freedesktop.org/software/systemd/man/systemd.unit.html#:~:text=The%20total%20length%20of%20the,service%20%22%2C%20%22%20.

Reply:
I changed the limit to 256.